src/util/mpsc.rs

```diff
@@ -77,97 +76,115 @@ impl<T> Clone for Sender<T> {
 impl<T> Drop for Sender<T> {
     #[inline]
     fn drop(&mut self) {
-        self.state.sender_cnt.fetch_add(-1, Ordering::AcqRel);
-        if self.state.is_sender_closed() {
-            self.notify();
+        let res = self.state.sender_cnt.fetch_add(-1, Ordering::AcqRel);
+        if res == 1 {
+            self.close_sender();
         }
     }
 }
 
+/// The receive end of a channel.
 pub struct Receiver<T> {
-    receiver: crossbeam_channel::Receiver<T>,
+    receiver: channel::Receiver<T>,
     state: Arc<State>,
 }
 
 impl<T> Sender<T> {
+    /// Returns the number of messages in the channel.
     #[inline]
-    fn notify(&self) {
-        let task = self.state.recv_task.take(Ordering::AcqRel);
-        if let Some(t) = task {
-            t.notify();
-        }
+    pub fn len(&self) -> usize {
+        self.sender.len()
     }
 
+    /// Returns true if the channel is empty.
+    ///
+    /// Note: Zero-capacity channels are always empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.sender.is_empty()
+    }
+
+    /// Blocks the current thread until a message is sent or the channel is disconnected.
     #[inline]
     pub fn send(&self, t: T) -> Result<(), SendError<T>> {
-        if !self.state.is_receiver_closed() {
-            self.sender.send(t);
-            self.notify();
-            return Ok(());
+        if self.state.is_sender_connected() {
```

Comment:
The latest `crossbeam::channel::Sender::send` returns `Result` now, so how about remove the `state` at all?

Reply:
`State` is still useful for close sender.

Reply:
Got it.